fs/move.py

```diff
@@ -55,24 +59,57 @@ def move_file(
             resources (defaults to `False`).
 
     """
-    with manage_fs(src_fs) as _src_fs:
-        with manage_fs(dst_fs, create=True) as _dst_fs:
+    with manage_fs(src_fs, writeable=True) as _src_fs:
+        with manage_fs(dst_fs, writeable=True, create=True) as _dst_fs:
             if _src_fs is _dst_fs:
                 # Same filesystem, may be optimized
                 _src_fs.move(
                     src_path, dst_path, overwrite=True, preserve_time=preserve_time
                 )
-            else:
-                # Standard copy and delete
-                with _src_fs.lock(), _dst_fs.lock():
-                    copy_file(
-                        _src_fs,
-                        src_path,
-                        _dst_fs,
-                        dst_path,
-                        preserve_time=preserve_time,
-                    )
+                return
+
+            if _src_fs.hassyspath(src_path) and _dst_fs.hassyspath(dst_path):
+                # if both filesystems have a syspath we create a new OSFS from a
+                # common parent folder and use it to move the file.
+
+                # we have to raise ResourceReadOnly manually if a FS is read-only
+                if _src_fs.getmeta().get("read_only", True):
+                    raise ResourceReadOnly(_src_fs, src_path)
+                if _dst_fs.getmeta().get("read_only", True):
+                    raise ResourceReadOnly(_dst_fs, dst_path)
+
+                try:
+                    src_syspath = _src_fs.getsyspath(src_path)
+                    dst_syspath = _dst_fs.getsyspath(dst_path)
+                    common = commonpath([src_syspath, dst_syspath])
+                    rel_src = frombase(common, src_syspath)
+                    rel_dst = frombase(common, dst_syspath)
+                    with _src_fs.lock(), _dst_fs.lock(), open_fs(
+                        common, writeable=True
+                    ) as base:
+                        base.move(rel_src, rel_dst, preserve_time=preserve_time)
+                    return  # optimization worked, exit early
+                except ValueError:
+                    # This is raised if we cannot find a common base folder.
+                    # In this case just fall through to the standard method.
+                    pass
+
+            # Standard copy and delete
+            with _src_fs.lock(), _dst_fs.lock():
+                copy_file(
+                    _src_fs,
+                    src_path,
+                    _dst_fs,
+                    dst_path,
+                    preserve_time=preserve_time,
+                )
+                try:
                     _src_fs.remove(src_path)
+                except FSError as e:
+                    # if the source cannot be removed we delete the copy on the
+                    # destination
+                    _dst_fs.remove(dst_path)
```

Comment:
Perhaps the better "fix" here is to just move your `.getmeta().get("read_only", True)` calls to be directly below the `manage_fs` lines?

Reply:
Can I depend on the fs being writeable if the meta info says so? What about read-only OSFS like a locked SD card or read-only linux mount?

Reply:
...or perhaps an "even better" fix is to tweak the behaviour of `manage_fs`, as I just suggested in another comment :wink:

Reply:
> Can I depend on the fs being writeable if the meta info says so? What about read-only OSFS like a locked SD card or read-only linux mount?

Good point - perhaps we need to add an extra keyword-arg so that the user can choose the desired behaviour if it turns out not to be possible to remove the original source copy?

Reply:
It's still the question whether the destination should be cleaned up in case of errors

Reply:
@tfeldmann :point_up: in case you missed this comment / question :wink:

Reply:
Just had a thought... we probably want the same `cleanup_dest_on_error` parameter for both `move_file` and `move_dir`?
And in the `move_dir` case, if we've moved files `a`, `b` and `c` from `src` to `dest` (and then deleted them from `src`), but we run into an error in transferring `d` (e.g. it's a network filesystem and the connection drops), we probably **don't** want to then delete `a`, `b` and `c` from `dest`?

So I guess we could either have `move_dir` defaulting `cleanup_dest_on_error` to `False`, and `move_file` defaulting it to `True`; or maybe it's more consistent to default it to `False` for both functions? :thinking: :man_shrugging:  What do the other @PyFilesystem/maintainers think?

Reply:
I was just in the middle of implementing this and had the same thought. Another option is to track which files were copied over and just clean up those?

Reply:
Cleaning up after `move_dir` could get quite complicated. Suppose you copy into a folder already containing other files, the cleanup should in no case just use `removetree`.

Reply:
> Suppose you copy into a folder already containing other files...

Urgh, good point - perhaps that's a good enough reason to go for "_not_ have a `cleanup_dest_on_error` parameter for `move_dir`" :slightly_smiling_face: 

Although it does prompt an interesting question - if the `dest_fs` _already_ had a `fileA`, and you did `move_file(src_fs, 'fileA', dest_fs, 'fileA')`, would deleting `fileA` from `dest_fs` still be the right thing to do if deleting `fileA` from `src_fs` failed? :thinking:  (IYSWIM)